[SPARK-9325][SPARK-R] head() and show() for Columns

olarayej opened the pull request:
# See attached design document

[SparkR collect (JIRA doc).pdf](https://github.com/apache/spark/files/143656/SparkR.collect.JIRA.doc.pdf)

Commit:
Support for collect() on Columns

Commit:
Removed drop=F from other PR

SparkQA commented:
**[Test build #51820 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51820/consoleFull)** for PR 11336 at commit [`fbf9b02`](https://github.com/apache/spark/commit/fbf9b02b478b8eb4845232e09932d068cb393fd8).

SparkQA commented:
**[Test build #51820 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51820/consoleFull)** for PR 11336 at commit [`fbf9b02`](https://github.com/apache/spark/commit/fbf9b02b478b8eb4845232e09932d068cb393fd8).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51820/
Test FAILed.

falaki commented:
Nice! thanks for doing this.

olarayej commented:
Thanks @falaki!

olarayej commented:
Jenkins, retest please. All tests pass for me after checking out this branch.

Commit:
Removed NA default value for df in column() method

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51832/
Test FAILed.

olarayej commented:
@AmplabJenkins Jenkins, could you retest please? I see ERROR: Error fetching remote repo 'origin'

Commit:
Added docs to head(Column)

SparkQA commented:
**[Test build #51844 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51844/consoleFull)** for PR 11336 at commit [`2d9ee18`](https://github.com/apache/spark/commit/2d9ee18dbdbcd25882a18e8fa0fa404d87c497c1).

SparkQA commented:
**[Test build #51844 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51844/consoleFull)** for PR 11336 at commit [`2d9ee18`](https://github.com/apache/spark/commit/2d9ee18dbdbcd25882a18e8fa0fa404d87c497c1).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51844/
Test FAILed.

Commit:
Still can't recreate build issue. Added default value df=NULL

SparkQA commented:
**[Test build #51846 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51846/consoleFull)** for PR 11336 at commit [`dc3df19`](https://github.com/apache/spark/commit/dc3df1926bb3931459a0ea381e5695b7bb56ca15).

SparkQA commented:
**[Test build #51846 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51846/consoleFull)** for PR 11336 at commit [`dc3df19`](https://github.com/apache/spark/commit/dc3df1926bb3931459a0ea381e5695b7bb56ca15).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51846/
Test FAILed.

Commit:
Noticed collate order is automatically changed after running install-dev.sh. Therefore, adding DESCRIPTION

SparkQA commented:
**[Test build #51848 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51848/consoleFull)** for PR 11336 at commit [`1e06d3c`](https://github.com/apache/spark/commit/1e06d3ccbb0cc441b9929f626812b8bc7dcbbda3).

SparkQA commented:
**[Test build #51848 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51848/consoleFull)** for PR 11336 at commit [`1e06d3c`](https://github.com/apache/spark/commit/1e06d3ccbb0cc441b9929f626812b8bc7dcbbda3).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51848/
Test FAILed.

Commit:
Still can't recreate build issue. Added minimal test case to debug it

SparkQA commented:
**[Test build #51860 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51860/consoleFull)** for PR 11336 at commit [`d697d44`](https://github.com/apache/spark/commit/d697d447ab8467eedade5d6e4c130dcf9a251e5e).

SparkQA commented:
**[Test build #51860 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51860/consoleFull)** for PR 11336 at commit [`d697d44`](https://github.com/apache/spark/commit/d697d447ab8467eedade5d6e4c130dcf9a251e5e).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51860/
Test FAILed.

olarayej commented:
@sun-rui @shivaram Do you know which version of R and SparkR's dependencies are being used in Jenkins? Tests run fine in my environment (have reviewed my code and ran unit tests many times). Wondering if that's due to a different version of R, testthat, devtools, rJava, etc.

Commit:
Reverted removing spaces in test cases. This works in my environment, though. However, it fails on Jenkins

SparkQA commented:
**[Test build #51863 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51863/consoleFull)** for PR 11336 at commit [`24b6154`](https://github.com/apache/spark/commit/24b6154b127ca57e17fa79831b2fee1edcc38b1b).

SparkQA commented:
**[Test build #51863 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51863/consoleFull)** for PR 11336 at commit [`24b6154`](https://github.com/apache/spark/commit/24b6154b127ca57e17fa79831b2fee1edcc38b1b).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51863/
Test FAILed.

sun-rui commented:
The R version used by Jenkins seems to be 3.2.x (can't remember clearly). There is a JIRA requesting using R 3.1.1, https://issues.apache.org/jira/browse/SPARK-11255, but it is not fixed.

felixcheung commented:
It was changed to 3.1.1 actually but there is a request to change to 3.1.2 hence the JIRA was reopened.






On Wed, Feb 24, 2016 at 3:43 AM -0800, "sun-rui" <notifications@github.com> wrote:





The R version used by Jenkins seems to be 3.2.x (can't remember clearly). There is a JIRA requesting using R 3.1.1, https://issues.apache.org/jira/browse/SPARK-11255, but it is not fixed.

---
Reply to this email directly or view it on GitHub:
https://github.com/apache/spark/pull/11336#issuecomment-188214706

shivaram commented:
Looks like the last one was just a flaky test in mllib. Lets try again 

Jenkins, retest this please

Commit:
Added docs for collect(Column)

SparkQA commented:
**[Test build #51904 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51904/consoleFull)** for PR 11336 at commit [`6a38a3c`](https://github.com/apache/spark/commit/6a38a3ca79087538dd7d3452a3cb487f9b0b6dd2).

SparkQA commented:
**[Test build #51904 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51904/consoleFull)** for PR 11336 at commit [`6a38a3c`](https://github.com/apache/spark/commit/6a38a3ca79087538dd7d3452a3cb487f9b0b6dd2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51904/
Test PASSed.

olarayej commented:
Thanks, folks. Looks like all test pass now! :-)

However, on my environment (R 3.2.2), two tests don't pass. We should be careful whenever upgrading the R version:

```
1. Failure (at test_sparkSQL.R#1052): column functions -------------------------
result not equal to expected
Names: 1 string mismatch

2. Failure (at test_sparkSQL.R#1058): column functions -------------------------
result not equal to expected
Names: 1 string mismatch
Error: Test failures
```

felixcheung commented:
The office [document](http://spark.apache.org/docs/latest/index.html) says Spark supports `R 3.1+`, so it is reasonable to assume R 3.2.2 should work with SparkR as well. I run R 3.2.2.

Is there a way to make it work with 3.1.x and 3.2.x?

olarayej commented:
@felixcheung It wasn't an R issue after all. The problem was that I hadn't been able to rebuild Spark in the last couple days due to SPARK-13431, and I needed changes on SPARK-12799. Now that it’s fixed, everything runs fine on R 3.2.2. Thank you!

olarayej commented:
Can any of you folks please take a look at the code? Thanks!

sun-rui commented:
@olarayej, I am not sure if it is conceptually correct to associate a Column to only one DF. Conceptually, a Column could be depend on 0, 1, 2 or more DataFrames. For example:
c1 <- df1$c1
c2 <- df2$c2
c3 < - c1 + c2

olarayej commented:
SparkR doesn't support operations between columns from different DataFrame objects. Yet you can do:

```
c1 <- df1$c1
c2 <- df2$c2
c3 < - c1 + c2
```
c3 can't be used at all. See examples below:

```
## Create two DataFrames from Iris
> irisDF <- createDataFrame(sqlContext, iris)
> irisDF2 <- createDataFrame(sqlContext, iris)

## Create Column x, adding two Columns in two DataFrame's
> x <- irisDF$Sepal_Length + irisDF2$Sepal_Length

## You can't use Column x as a predicate
> irisDF[x > 0, ]
16/03/01 11:04:19 ERROR RBackendHandler: filter on 76 failed
Error in invokeJava(isStatic = FALSE, objId$id, methodName, ...) : 
  org.apache.spark.sql.AnalysisException: resolved attribute(s) Sepal_Length#20 missing from Sepal_Length#15,Petal_Width#18,Sepal_Width#16,Petal_Length#17,Species#19 in operator !Filter ((Sepal_Length#15 + Sepal_Length#20) > 0.0);
	
## You can't select Column x either
> select(irisDF, x)
16/03/01 11:04:43 ERROR RBackendHandler: select on 76 failed
Error in invokeJava(isStatic = FALSE, objId$id, methodName, ...) : 
  org.apache.spark.sql.AnalysisException: resolved attribute(s) Sepal_Length#20 missing from Sepal_Length#15,Petal_Width#18,Sepal_Width#16,Petal_Length#17,Species#19 in operator !Project [(Sepal_Length#15 + Sepal_Length#20) AS (Sepal_Length + Sepal_Length)#25];

> select(irisDF2, x)
16/03/01 11:04:45 ERROR RBackendHandler: select on 91 failed
Error in invokeJava(isStatic = FALSE, objId$id, methodName, ...) : 
  org.apache.spark.sql.AnalysisException: resolved attribute(s) Sepal_Length#15 missing from Sepal_Length#20,Sepal_Width#21,Species#24,Petal_Width#23,Petal_Length#22 in operator !Project [(Sepal_Length#15 + Sepal_Length#20) AS (Sepal_Length + Sepal_Length)#26];

```
In my opinion, we should throw an error if the user is trying to operate on Columns coming from different DataFrames.

sun-rui commented:
@olarayej, 
c3 can be used on a DataFrame that is joined between df1 & df2
df3 <- join(df1, df2)

olarayej commented:
@sun-rui Yes. In that case, c3 will be only associated to df3.

olarayej commented:
Also, the fact that the size of a column depends on the join seems counter-intuitive for an R user:

```
> dim(irisDF2)
[1] 150   5

> dim(irisDF)
[1] 150   5

> x <- irisDF$Sepal_Length + irisDF2$Sepal_Length
```
In R, x will always have 150 elements. However:

```
# Cartesian product
> df3 <- join(irisDF, irisDF2)
> dim(select(df3, x))
[1] 22500     1

# Inner join by Species
> df4 <- merge(irisDF, irisDF2, by="Species")
> dim(select(df4, x))
[1] 7500    1

```
I still think SparkR shouldn't allow operations between columns coming from different DataFrames. And, in the case of a join, operations can be performed on the joined DataFrame (e.g., df3) as opposed to the original ones (e.g., irisDF and irisDF2).

olarayej commented:
@sun-rui Does that make sense to you? @shivaram @felixcheung Any comments?

sun-rui commented:
A column can be applied to different dataframes.
For example, if both df1 and df2 have a column named "col",then
col <- column("col")
collect(select(df1, col))
collect(select(df2, col))
both works.

Take the join case above as example,
You have can different DataFrames resulting from different joins on both df1 and df2,
and apply c3 to the different resulting DataFrames also work.

So how do you know which dataFrame to associate with a column in such cases?

@rxin, any comments on this issue?

rxin commented:
I personally find it confusing having to reason about when we can "head"/"collect"/"show" and when we cannot, and that's why the Scala/Python version of the API didn't have this feature.

aloknsingh commented:
Hi @rxin ,
If we see that in general R have more features than python and other language as far as stats and ml is concerned. So it would be nice to have more R like approach than having the global goal of compatibility in all the language support. so as long as JavaDataFrame is compatible in all the language one can always add more functions on the language specific dataframe i.e in this case R_DataFrame. Even though these features are not the main goal of SparkR , I think R user will find it more R like and appreciate it. Just come thoughts :)

shivaram commented:
Yeah I'm not sure whats the right resolution on this -- On the one hand I have to say that the Spark DataFrame model where a column doesn't belong to one (or any) DataFrame is more confusing from a users perspective.  

However the risk of adding wrappers which don't follow the underlying data model is that it might lead to inconsistencies where some parts of the API don't work like others.

Commit:
Handled case of 'orhpan' Columns with no parent DataFrame

Commit:
Added tests for orphan Columns

SparkQA commented:
**[Test build #54546 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54546/consoleFull)** for PR 11336 at commit [`c86bebb`](https://github.com/apache/spark/commit/c86bebb9ce02e5623d328d33109fff1ee8eeca56).

SparkQA commented:
**[Test build #54546 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54546/consoleFull)** for PR 11336 at commit [`c86bebb`](https://github.com/apache/spark/commit/c86bebb9ce02e5623d328d33109fff1ee8eeca56).
 * This patch **fails R style tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54546/
Test FAILed.

SparkQA commented:
**[Test build #54557 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54557/consoleFull)** for PR 11336 at commit [`bf6c456`](https://github.com/apache/spark/commit/bf6c45601a83d602853cce359d1d040cd36b1d8d).

Commit:
Merged with the base branch

SparkQA commented:
**[Test build #54557 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54557/consoleFull)** for PR 11336 at commit [`bf6c456`](https://github.com/apache/spark/commit/bf6c45601a83d602853cce359d1d040cd36b1d8d).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54557/
Test FAILed.

Commit:
Fixed style issues

SparkQA commented:
**[Test build #54559 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54559/consoleFull)** for PR 11336 at commit [`e5659ee`](https://github.com/apache/spark/commit/e5659ee85391937cdf301bd1acc01487bc55c129).

SparkQA commented:
**[Test build #54559 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54559/consoleFull)** for PR 11336 at commit [`e5659ee`](https://github.com/apache/spark/commit/e5659ee85391937cdf301bd1acc01487bc55c129).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54559/
Test PASSed.

olarayej commented:
Thanks @sun-rui @rxin @shivaram  for your inputs. To alleviate the confusion on which columns can/cannot be collected, I propose the following (already pushed the code):

Currently there are 15 SparkR functions that return an ‘orphan’ Column with no parent DataFrame:
```
rand, rand, unix_timestamp,
struct, expr, column, lag, lead, lit, cume_dist, dense_rank,
ntile, percent_rank, rank, row_number
```
The first three (i.e., rand, randn, and unix_timestamp) can be nicely collected as single elements. For example:
```
> rand()
[1] 0.01483325
```
The remaining ones don’t make sense unless there’s an associated DataFrame. Therefore, an empty vector will be returned:
```
> column("Species")
Species
<Empty column>

> collect(column("Species"))
character(0)
```

I think it makes sense: If you don’t associate a Column with a DataFrame, there’s nothing to be collected. Now, for Columns that do belong to a DataFrame, collecting columns SIGNIFICANTLY improves usability in 138 functions/operators (besides other issues in the design document), for example:

> irisDF$Sepal_Length * 100
 [1] 510 490 470 460 500 540 460 500 440 490 540 480 480 430 580 570 540 510 570 510…

versus:

> head(select(irisDF, irisDF$Sepal_Length * 100), 20)[, 1]
 [1] 510 490 470 460 500 540 460 500 440 490 540 480 480 430 580 570 540 510 570 510

@shivaram has a very valid point: this introduces discrepancies in the Spark API’s across multiple languages. I believe this is not necessarily bad as R, especially, is a slightly different animal which already has a specific behavior for columns (i.e., vectors).

olarayej commented:
@shivaram @falaki @felixcheung @rxin @sun-rui Any thoughts on this?

shivaram commented:
Thanks @olarayej for the clarifications - Just broadening the discussion a bit:
@mengxr @thunterdb Could you comment if this will be useful in some of the ML use cases you are working on ?

felixcheung commented on R/pkg/R/column.R:
```diff
@@ -14,6 +14,8 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 #
+setOldClass("DataFrame")
```
Is this necessary - setClass has been called on DataFrame?

SparkQA commented:
**[Test build #58038 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58038/consoleFull)** for PR 11336 at commit [`9c1661f`](https://github.com/apache/spark/commit/9c1661f0fe892fd3dcd40e8e489a940c9b129747).

Commit:
Merged upstream branch

Commit:
pkg/R/columnR

Commit:
Merge branch 'master' of https://github.com/apache/spark into SPARK-9325

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58038/
Test PASSed.

olarayej commented:
I just got this branch up to date. Any comments, folks? @shivaram @falaki @felixcheung @rxin @sun-rui @mengxr

sun-rui commented:
This tends to use a Column as a distributed vector :)  I don't object to add R-like features in SparkR, but am still not sure on this feature. I am neutral and up to other committer's vote

falaki commented:
@olarayej are you interested in rebasing this PR and limiting its user impact to just `head` for now? We can continue the discussion on support for distributed vector (a.k.a. `Column`) in JIRA and when resolved create follow up PRs.

olarayej commented:
@falaki Absolutely. Let me do that. Thank you!

falaki commented:
@olarayej any update not his. If you are busy I can start another PR from yours.

olarayej commented:
@falaki Sorry, I was out of town. Let me get back to this today. Thank you!

SparkQA commented:
**[Test build #66753 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66753/consoleFull)** for PR 11336 at commit [`8f906a2`](https://github.com/apache/spark/commit/8f906a2e8050c391355c3ddab811d990020728cb).

Commit:
Merged with upstream

SparkQA commented:
**[Test build #66753 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66753/consoleFull)** for PR 11336 at commit [`8f906a2`](https://github.com/apache/spark/commit/8f906a2e8050c391355c3ddab811d990020728cb).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66753/
Test FAILed.

falaki commented on R/pkg/R/DataFrame.R:
```diff
@@ -1035,10 +1035,17 @@ setMethod("dim",
 #'
-#' @param x a SparkDataFrame.
-#' @param stringsAsFactors (Optional) a logical indicating whether or not string columns
+#' If applied to a SparkDataFrame, \code{collect} returns a data.frame. If applied to a 
```
Do we still want this documentation?

falaki commented on R/pkg/R/column.R:
```diff
@@ -32,35 +34,65 @@ setOldClass("jobj")
+  }
+})
+
+setMethod("collect", signature = "Column", definition = function(x) {
```
It seems Spark committers are still not sure about this feature. Maybe we can remove it for now and once the discussion is resolved we work on it as a separate PR?

falaki commented on R/pkg/R/column.R:
```diff
@@ -32,35 +34,65 @@ setOldClass("jobj")
   .Object
 })
 
+setMethod("show", signature = "Column", definition = function(object) {
```
Maybe this can be implemented as `head` with default number of elements?

felixcheung replied:
+1, default to 6 for consistency?

olarayej replied:
@falaki @felixcheung I remember discussing this issue in the past when working on PR 8913. Adding parameter `n` to `show()` brings extra complications as follows:

Since parameter `n` is not part of `show()` signature, I'll need to change the generic definition, adding `... ` as an additional parameter. This causes some inconsistencies:

```
> df$Sepal_Length
An object of class "Column"
Slot "jc":
Java ref type org.apache.spark.sql.Column id 18 

Slot "df":
SparkDataFrame[Sepal_Length:double, Sepal_Width:double, Petal_Length:double, Petal_Width:double, Species:string]
```

```
> show(df$Sepal_Length)
 [1] 5.1 4.9 4.7 4.6 5.0 5.4 4.6 5.0 4.4 4.9 5.4 4.8 4.8 4.3 5.8 5.7 5.4 5.1 5.7
[20] 5.1
```
In the first case, the method that's invoked is `showDefault()` which, in turn invokes `show()`, but since I have changed the definition of `show()`, it is dispatching `methods:::show()` as opposed to `SparkR:::show()`. To fix this, I'll have to make `showDefault()` a generic and re-implement it, which I'd rather not to avoid potential conflicts with base R functions.

My suggestion would be to have a global parameter for the number of rows/elements to show which is **exactly** what R does.

falaki commented on R/pkg/inst/tests/testthat/test_sparkSQL.R:
```diff
@@ -2252,6 +2252,31 @@ test_that("Method str()", {
   expect_equal(capture.output(utils:::str(iris)), capture.output(str(iris)))
 })
 
+test_that("collect/show/head on Columns", {
```
I think if we remove `collect()` from this PR the test needs to be updated as well.

olarayej commented:
@falaki Thanks for your comments. Yeah, before removing collect/show, I just wanted to rebase to current upstream master. I'm getting a build error which is actually a warning, not even an R error:

```
(converted from warning) 'sparkRSQL.init' is deprecated.
Use 'sparkR.session' instead.
```

I don't explicitly use sparkRSQL.init anywhere in my code so I'm investigating. If you have any suggestion, it'd be nice. Thanks!

falaki commented:
I think some of the old tests still rely on `sparkRSQL.init()`. I believe the warning is OK.

olarayej commented:
@falaki Yeah but those warnings are making the build fail (see below). Is that okay? Now I see a new"Checks" section. I may be outdated with the protocols as it's been a while I didn't commit :-). Thanks!

```
Failed -------------------------------------------------------------------------
1. Error: column binary mathfunctions (@test_sparkSQL.R#1256) ------------------
error in evaluating the argument 'x' in selecting a method for function 'collect': 
  error in evaluating the argument 'col' in selecting a method for function 'select': (converted from warning) 'sparkRSQL.init' is deprecated.
Use 'sparkR.session' instead.
See help("Deprecated")
1: expect_equal(class(collect(select(df, rand()))[2, 1]), "numeric") at /home/jenkins/workspace/SparkPullRequestBuilder/R/lib/SparkR/tests/testthat/test_sparkSQL.R:1256
2: compare(object, expected, ...)
3: collect(select(df, rand()))
```

SparkQA commented:
**[Test build #66767 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66767/consoleFull)** for PR 11336 at commit [`ed0abf2`](https://github.com/apache/spark/commit/ed0abf24d7f65ad2381f6d664ba23e440013c97a).

Commit:
Removed method collect()

felixcheung commented on R/pkg/R/DataFrame.R:
```diff
@@ -1035,10 +1035,16 @@ setMethod("dim",
           })
 
-#' Collects all the elements of a SparkDataFrame and coerces them into an R data.frame.
+#' Download Spark datasets into R
```
I'm not sure this should say "datasets" - we don't have this term elsewhere

olarayej replied:
Sure. Thanks!

felixcheung commented on R/pkg/R/DataFrame.R:
```diff
@@ -1182,10 +1195,18 @@ setMethod("take",
-#' head(df)
+#' # Initialize Spark context and SQL context
+#' sc <- sparkR.init()
+#' sqlContext <- sparkRSQL.init(sc)
```
ditto here

felixcheung commented on R/pkg/R/DataFrame.R:
```diff
@@ -1168,12 +1179,14 @@ setMethod("take",
-#' @param x a SparkDataFrame.
-#' @param num the number of rows to return. Default is 6.
-#' @return A data.frame.
+#' @param x A SparkDataFrame or Column
```
for something like this the convention we have is to add the @param in generics.R - you can see other examples there

olarayej replied:
Not sure I follow here. Could you point to the specific example?

felixcheung commented on R/pkg/R/column.R:
```diff
@@ -32,35 +34,57 @@ setOldClass("jobj")
   .Object
 })
 
+setMethod("show", signature = "Column", definition = function(object) {
```
Could you check the style - we don't typically put "definition"

olarayej replied:
Sure

SparkQA commented:
**[Test build #66767 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66767/consoleFull)** for PR 11336 at commit [`ed0abf2`](https://github.com/apache/spark/commit/ed0abf24d7f65ad2381f6d664ba23e440013c97a).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66767/
Test FAILed.

felixcheung commented on R/pkg/R/functions.R:
```diff
@@ -2847,7 +2860,11 @@ setMethod("rand", signature(seed = "missing"),
+
+            # By assigning a one-row data.frame, the result of this function can be collected
+            # returning a one-element Column
+            df <- as.DataFrame(sparkRSQL.init(), data.frame(0))
```
ditto

felixcheung commented on R/pkg/R/functions.R:
```diff
@@ -2865,7 +2882,11 @@ setMethod("rand", signature(seed = "numeric"),
+
+            # By assigning a one-row data.frame, the result of this function can be collected
+            # returning a one-element Column
+            df <- as.DataFrame(sparkRSQL.init(), data.frame(0))
```
ditto

felixcheung commented on R/pkg/R/functions.R:
```diff
@@ -2876,7 +2897,8 @@ setMethod("randn", signature(seed = "missing"),
           function(seed) {
             jc <- callJStatic("org.apache.spark.sql.functions", "randn", as.integer(seed))
-            column(jc)
+            df <- as.DataFrame(sparkRSQL.init(), data.frame(0))
```
ditto

felixcheung commented on R/pkg/R/functions.R:
```diff
@@ -3026,7 +3048,11 @@ setMethod("translate",
+
+            # By assigning a one-row data.frame, the result of this function can be collected
+            # returning a one-element Column
+            df <- as.DataFrame(sparkRSQL.init(), data.frame(0))
```
ditto

felixcheung commented:
I know why tests fail - please see my comment.

olarayej commented on R/pkg/R/functions.R:
```diff
@@ -2836,7 +2845,11 @@ setMethod("lpad", signature(x = "Column", len = "numeric", pad = "character"),
+
+            # By assigning a one-row data.frame, the result of this function can be collected
+            # returning a one-element Column
+            df <- as.DataFrame(sparkRSQL.init(), data.frame(0))
```
See my comment from March 30 to illustrate why this is needed. I'll change sparkRSQL.init() to sparkR.session(). Thanks for catching this!

felixcheung replied:
actually, just change it to `as.DataFrame(data.frame(0))`

felixcheung replied:
in fact, it might be useful to have a shared empty DataFrame instead of instantiating a new one every time a SQL function is called (which could be quite frequent).

olarayej replied:
@felixcheung That's a good idea. I have created a singleton accordingly.

olarayej commented on R/pkg/R/DataFrame.R:
```diff
@@ -1049,11 +1055,16 @@ setMethod("dim",
-#' firstName <- collected[[1]]$name
+#' # Initialize Spark context and SQL context
+#' sc <- sparkR.init()
+#' sqlContext <- sparkRSQL.init(sc)
```
Sure. Thanks!

Commit:
Removed spark.init() call, fixed docs

SparkQA commented:
**[Test build #66832 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66832/consoleFull)** for PR 11336 at commit [`20e53e8`](https://github.com/apache/spark/commit/20e53e83a4d564c17d0b180ae5eab8ca9d6c1410).

SparkQA commented:
**[Test build #66832 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66832/consoleFull)** for PR 11336 at commit [`20e53e8`](https://github.com/apache/spark/commit/20e53e83a4d564c17d0b180ae5eab8ca9d6c1410).
 * This patch **fails some tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66832/
Test FAILed.

Commit:
Used a singleton to generate empty DataFrame

Commit:
Minor docs change

SparkQA commented:
**[Test build #66834 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66834/consoleFull)** for PR 11336 at commit [`266d5ff`](https://github.com/apache/spark/commit/266d5ffd87929adf081944de19ee36f892125977).

Commit:
Added missing bracket

SparkQA commented:
**[Test build #66834 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66834/consoleFull)** for PR 11336 at commit [`266d5ff`](https://github.com/apache/spark/commit/266d5ffd87929adf081944de19ee36f892125977).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

Commit:
Added documentation for parameter df of class Column

SparkQA commented:
**[Test build #66844 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66844/consoleFull)** for PR 11336 at commit [`257fa86`](https://github.com/apache/spark/commit/257fa8669eb0520d4827c0bc3d52a07037bd38f8).

SparkQA commented:
**[Test build #66844 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66844/consoleFull)** for PR 11336 at commit [`257fa86`](https://github.com/apache/spark/commit/257fa8669eb0520d4827c0bc3d52a07037bd38f8).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66844/
Test FAILed.

olarayej commented:
@falaki @felixcheung I have addressed all your comments. I'm getting two documentation warnings which seem to be making the build fail:

1)
```
Undocumented S4 methods:
  generic 'head' and siglist 'Column'
  generic 'show' and siglist 'Column'
```
The documentation for these is on Data.Frame.R. I don't see a need for duplicating the docs in column.R


2)
```
Undocumented arguments in documentation object 'Column-class'
  'df'

Undocumented arguments in documentation object 'head'
  '...'
```
I do have documentation for slot df in class Column. And also, I don't have ... as part of the signature of method head. Not sure why this warning comes up.

felixcheung commented on R/pkg/R/column.R:
```diff
@@ -29,38 +31,61 @@ setOldClass("jobj")
   .Object
 })
 
+setMethod("show", signature = "Column", function(object) {
```
you do need to add at least `#' @rdname show` here to tell roxygen2 they should go together

felixcheung commented on R/pkg/R/column.R:
```diff
@@ -29,38 +31,61 @@ setOldClass("jobj")
+  }
+})
+
+setMethod("head", signature = "Column", definition = function(x, n=6) {
```
ditto here, `#' @rdname head`

felixcheung commented:
@wangmiao1981 These are required for the recently added CRAN checks to ensure we could release SparkR as a R package.
1) - note in-line - they need to have `#' @rdname`
2) - I'm not sure about "Column-class" we should check the generated .Rd file
As for "head", I think it's because there is an existing generic and that has `...` in the signature
Did it work if you skip the generic here `setGeneric("head")`https://github.com/apache/spark/pull/11336/files#diff-8e3d61ff66c9ffcd6ffb7a8eedc08409R555
If not, we need to match the existing generic https://stat.ethz.ch/R-manual/R-devel/library/utils/html/head.html

SparkQA commented:
**[Test build #66905 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66905/consoleFull)** for PR 11336 at commit [`0691c32`](https://github.com/apache/spark/commit/0691c32cacc3218742c7f345b4bc498ba2f826e5).

Commit:
Fixed docs issues

SparkQA commented:
**[Test build #66905 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66905/consoleFull)** for PR 11336 at commit [`0691c32`](https://github.com/apache/spark/commit/0691c32cacc3218742c7f345b4bc498ba2f826e5).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66905/
Test FAILed.

SparkQA commented:
**[Test build #66911 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66911/consoleFull)** for PR 11336 at commit [`445407c`](https://github.com/apache/spark/commit/445407caa11b0f555e32d51008fdf83d5c4ecd97).

Commit:
Removed commented code

SparkQA commented:
**[Test build #66911 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66911/consoleFull)** for PR 11336 at commit [`445407c`](https://github.com/apache/spark/commit/445407caa11b0f555e32d51008fdf83d5c4ecd97).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66911/
Test FAILed.

SparkQA commented:
**[Test build #66921 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66921/consoleFull)** for PR 11336 at commit [`1ace2e5`](https://github.com/apache/spark/commit/1ace2e5785030df790b3565d2a0b24358229d655).

Commit:
Fixed docs issues. Renamed parameter n as num for method head to have consistency with DataFrame's head

SparkQA commented:
**[Test build #66921 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66921/consoleFull)** for PR 11336 at commit [`1ace2e5`](https://github.com/apache/spark/commit/1ace2e5785030df790b3565d2a0b24358229d655).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66921/
Test FAILed.

SparkQA commented:
**[Test build #66968 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66968/consoleFull)** for PR 11336 at commit [`2bfb8a6`](https://github.com/apache/spark/commit/2bfb8a62adf81e6953e398147e5f8d9122bc47ef).

Commit:
Style fixes

SparkQA commented:
**[Test build #66968 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66968/consoleFull)** for PR 11336 at commit [`2bfb8a6`](https://github.com/apache/spark/commit/2bfb8a62adf81e6953e398147e5f8d9122bc47ef).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66968/
Test FAILed.

SparkQA commented:
**[Test build #66973 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66973/consoleFull)** for PR 11336 at commit [`e0bba0a`](https://github.com/apache/spark/commit/e0bba0a0655f750751a6b9b05a7f6cf5eb807e05).

Commit:
Fixed docs issues

SparkQA commented:
**[Test build #66973 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66973/consoleFull)** for PR 11336 at commit [`e0bba0a`](https://github.com/apache/spark/commit/e0bba0a0655f750751a6b9b05a7f6cf5eb807e05).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66973/
Test PASSed.

olarayej commented:
@felixcheung @falaki I have addressed all your comments and tests pass now. Thank you! cc @aloknsingh

felixcheung commented:
could you update the JIRA and PR to reflect what's being added here?

felixcheung commented on R/pkg/R/DataFrame.R:
```diff
@@ -1166,26 +1166,33 @@ setMethod("take",
 #' @export
 #' @examples
 #'\dontrun{
+#' # Initialize Spark context and SQL context
```
I don't think we need this line. Also we are using Spark Session as the terminology.

felixcheung commented on R/pkg/R/DataFrame.R:
```diff
@@ -1166,26 +1166,33 @@ setMethod("take",
-#' head(df)
+#' 
+#' # Create a DataFrame from the Iris dataset
+#' irisDF <- as.DataFrame(iris)
```
you might want to avoid using `iris` in example, because it will cause a warning with the column name having `.` and if at one point we fix this to support `.` in column name, the example will need to be updated

felixcheung commented on R/pkg/R/DataFrame.R:
```diff
@@ -3321,3 +3328,11 @@ setMethod("randomSplit",
+  if (is.null(.sparkREnv$EMPTY_DF)) {
+    .sparkREnv$EMPTY_DF <- as.DataFrame(data.frame(0))
+  }
+  return(.sparkREnv$EMPTY_DF)
```
nite, style - don't need `return`

felixcheung replied:
also, the convention seems to be `get(".sparkRsession", envir = .sparkREnv)` - with the object name in `.lowerCamalCase` - could you change to that

olarayej replied:
get() would throw an error if the variable is not defined. I'll use exists()

felixcheung commented on R/pkg/R/column.R:
```diff
@@ -29,38 +31,66 @@ setOldClass("jobj")
+          signature = "Column",
+          function(x, num = 6L) {
+            if (is.null(x@df)) {
+              collect(x)
```
we don't have a `collect` method for Column right? Does this work?

felixcheung replied:
so this means this here should be changed, or...?

Commit:
Cosmetic changes

felixcheung commented on R/pkg/R/column.R:
```diff
@@ -29,38 +31,66 @@ setOldClass("jobj")
+
+#' @rdname head
+setMethod("head",
+          signature = "Column",
```
also, this form is a bit implicit, could you update these signatures to say something like `signature(x = "jobj")` with the variable names

SparkQA commented:
**[Test build #67220 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67220/consoleFull)** for PR 11336 at commit [`76061ad`](https://github.com/apache/spark/commit/76061ad8320c0ab1a52da1d0a7ea96e2db5d8a68).

Commit:
More cosmetics

SparkQA commented:
**[Test build #67220 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67220/consoleFull)** for PR 11336 at commit [`76061ad`](https://github.com/apache/spark/commit/76061ad8320c0ab1a52da1d0a7ea96e2db5d8a68).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67220/
Test PASSed.

Commit:
More cosmetics

SparkQA commented:
**[Test build #67348 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67348/consoleFull)** for PR 11336 at commit [`ed1b382`](https://github.com/apache/spark/commit/ed1b382524265340d0b692a89c6be388eddb549a).

SparkQA commented:
**[Test build #67348 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67348/consoleFull)** for PR 11336 at commit [`ed1b382`](https://github.com/apache/spark/commit/ed1b382524265340d0b692a89c6be388eddb549a).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67348/
Test FAILed.

SparkQA commented:
**[Test build #67351 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67351/consoleFull)** for PR 11336 at commit [`1338d71`](https://github.com/apache/spark/commit/1338d71a8f8f1387e08326851e15fcc84d5092f9).

Commit:
Fixed style issues

SparkQA commented:
**[Test build #67351 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67351/consoleFull)** for PR 11336 at commit [`1338d71`](https://github.com/apache/spark/commit/1338d71a8f8f1387e08326851e15fcc84d5092f9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

olarayej commented:
@felixcheung @falaki I have addressed all your comments. Shall we merge?

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67351/
Test PASSed.

olarayej commented:
@felixcheung @falaki Folks?

felixcheung commented:
do you have tests for the case where the df is null, like https://github.com/apache/spark/pull/11336/files#diff-04c14efaae2b7b0f0a45038482f2590cR77?
There are several other cases too.

felixcheung commented:
one thing to highlight, previously we discussed a proposal to display a subset of values for a variable, such as when typing in the variable name in R shell or RStudio

```
> a <- iris
> a
    Sepal.Length Sepal.Width Petal.Length Petal.Width    Species
1            5.1         3.5          1.4         0.2     setosa
2            4.9         3.0          1.4         0.2     setosa
3            4.7         3.2          1.3         0.2     setosa
```

This PR includes a change to how Column is presented in this case,

before
```
> a <- as.DataFrame(iris)
> a
SparkDataFrame[Sepal_Length:double, Sepal_Width:double, Petal_Length:double, Petal_Width:double, Species:string]
> a$Sepal_Length
Column Sepal_Length
```

after
```
> a$Sepal.Length
  [1] 5.1 4.9 4.7 4.6 5.0 5.4 4.6 5.0 4.4 4.9 5.4 4.8 4.8 4.3 5.8 5.7 5.4 5.1
 [19] 5.7 5.1 5.4 5.1 4.6 5.1 4.8 5.0 5.0 5.2 5.2 4.7 4.8 5.4 5.2 5.5 4.9 5.0
```

So instead of type and name we would be doing a collect.

In the earlier discussion, it was concluded that we would not be doing this without an explicit method call (like `head(col)`), unless, perhaps when the data frame is local.

olarayej commented:
@felixcheung When a user types a variable name on the R shell, it triggers method showDefault() which, in turn, invokes show(). I wrote an implementation of show() for Column which, in turn, invokes head() (not collect), showing the first 20 elements of the dataset. This mimics R behavior and I think it also helps with usability. However, if the agreement is not to have that, I can just remove show method.

felixcheung commented:
Just pointing out what might not be obvious for other observers.
:)

Commit:
Added test for Columns with no parent DataFrame

Commit:
Moved head documentation to generics.R

SparkQA commented:
**[Test build #67536 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67536/consoleFull)** for PR 11336 at commit [`619f23b`](https://github.com/apache/spark/commit/619f23b367b94a47f3f63dea008ab4d674218b34).

Commit:
Moved documentation for show method to generics.R

SparkQA commented:
**[Test build #67536 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67536/consoleFull)** for PR 11336 at commit [`619f23b`](https://github.com/apache/spark/commit/619f23b367b94a47f3f63dea008ab4d674218b34).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67536/
Test PASSed.

olarayej commented:
@felixcheung I'm done! Thanks for your comments!

shivaram commented:
@olarayej @felixcheung Sorry I've missed the updates to this. I'll try to take a look at this later today ?

olarayej commented:
@shivaram @falaki @felixcheung Any additional comments? Otherwise, are we ready to merge?

olarayej commented:
Folks?

olarayej commented:
@shivaram @felixcheung @falaki Any thoughts?

falaki commented:
I did another pass. It looks good to me.

olarayej commented:
@falaki @shivaram Shall we merge this?

felixcheung commented:
I think @shivaram should review this.

shivaram commented:
Sorry this got dropped from my radar as I was caught up with some other stuff. I am more free for the next 2-3 days. Will review this

olarayej commented:
Happy New Year, folks! Any updates on this? @shivaram @falaki

gatorsmile commented:
Any update on this PR?

gatorsmile commented:
ping @olarayej 

cc @felixcheung

olarayej commented:
@shivaram: Have you reviewed this? If the intent is to merge it, I'll gladly update the code. @gatorsmile

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/4296/
Test PASSed.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/4329/
Test PASSed.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/4336/
Test PASSed.

SparkQA commented:
**[Test build #97866 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97866/consoleFull)** for PR 11336 at commit [`619f23b`](https://github.com/apache/spark/commit/619f23b367b94a47f3f63dea008ab4d674218b34).

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/4358/
Test PASSed.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/15824/
Test PASSed.
